File: 1228/c/c.py
mod = 1000000007

def sqrat(n):
    x = n
    y = (x+1) // 2
    while y < x:
        x = y
        y = (x + n // x) // 2
    return x

def paw(e, exp):
    if exp == 0: return 1
    if exp == 1: return e
    ret = paw(e, exp//2);
    l = ret * ret % mod
    if exp%2 == 1:
        l *= e
        l %= mod
    return l

def main():
    inp = input().split()
    x = int(inp[0])
    n = int(inp[1])
    prime = set()

    cur = x;
    for i in range(2,sqrat(x)+5):
        while cur % i == 0:
            cur //= i
            prime.add(i)
    if cur > 1:
        prime.add(cur)

    p = 1
    for e in prime:
        f = e
        exp = 0
        while f <= n:
            exp += n //f
            f *= e
        p *= paw(e, exp)
        p %= mod

    print(p)

main()

File: 1228/c/c.cpp
// youngjinp20
// 2019 09

#include <bits/stdc++.h>
using namespace std;

typedef unsigned long long ll;

ll mod = 1000000007;

ll paw(int e, ll exp) {
	if (exp == 0) return 1;
	if (exp == 1) return e;
	ll ret = paw(e, exp/2);
	ll l = ret * ret;
	l %= mod;
	l *= paw(e, exp-2*(exp/2));
	l %= mod;
	return l;
}

int main() {
	int       x;
	ll n;
	set<int>  prime;

	cin >> x >> n;

	// Get prime numbers in set
	int cur = x;
	for (int i=2; i<sqrt(x)+5; i++) {
		while (cur%i == 0) {
			cur /= i;
			prime.insert(i);
		}
	}
	if (cur > 1) prime.insert(cur);

	// Other shit
	ll p = 1;
	for (auto e : prime) {
		ll f = e;
		ll exp = 0;
		while (f <= n) {
			exp += n / f;
			if (f <= ULONG_MAX / e) f *= e;
			else break;
		}
		p *= paw(e, exp);
		p %= mod;
	}

	printf("%lld", p);

	return 0;
}
